File: scripts/definir_ultimo_precio.py
import mysql.connector
from datetime import datetime, timedelta

conexion = mysql.connector.connect(
    host="localhost",
    user="precios",
    password="precios",
    database="precios"
)

cursor = conexion.cursor()

cursor.execute("SELECT * FROM products")
resultados = cursor.fetchall()

for producto in resultados:
    print(producto)
    id_producto = producto[0]
    print(id_producto)

    cursor.execute("SELECT * FROM price WHERE product_id = %s ORDER BY date_time DESC LIMIT 1", (id_producto,))
    ultimo_precio = cursor.fetchone()
    
    if ultimo_precio != None:
        valor_ultimo_precio = ultimo_precio[2]
        fecha_ultimo_precio = ultimo_precio[3]
        
        print('Fecha ultimo precio',fecha_ultimo_precio)
        print("Precio: ", valor_ultimo_precio)

        sql = "UPDATE products SET ultimo_precio_conocido = %s WHERE id = %s"
        cursor.execute(sql, (fecha_ultimo_precio, id_producto,))
        sql = "UPDATE products SET last_price = %s WHERE id = %s"
        cursor.execute(sql, (valor_ultimo_precio, id_producto,))

conexion.commit()

File: scripts/definir_ultimo_precio.js
require("dotenv").config({ path: '../.env' })
const { text } = require("express");
const fs = require('fs');
const { resolve } = require("path");
const uuid = require("uuid")

let conn_obj = {
    host: process.env.mysql_host,
  //  port: process.env.mysql_port,
    user: process.env.mysql_user,
    password: process.env.mysql_password,
    database: process.env.mysql_database,
    supportBigNumbers: true,
    bigNumberStrings: true,
    typeCast: function (field, next) {
        if (field.type == "NEWDECIMAL") {
            var value = field.string();
            return (value === null) ? null : Number(value);
        }
        return next();
    }
  
  }
  
const knex = require('knex')({
    client: 'mysql2',
    connection: conn_obj,
    pool: { min: 0, max: 1000, "propagateCreateError": false }
})

let i=0
async function definir_ultimo( producto ){
    let trx = await knex.transaction()
    console.log(i++,' - ','procesando ', producto.name, '  ',producto.id )
    let precios = await knex.raw(`
        SELECT *
            FROM price
            WHERE product_id = :prod_id
            AND date_time = (
            SELECT MAX(date_time)
            FROM price
            WHERE product_id = :prod_id
            AND branch_id = price.branch_id
        )
    `,{ prod_id:producto.id})
    if (precios){ 
        precios = precios[0]
        console.log(precios)
        let proms_ = []
        for (let i=0; i < precios.length; i++){
            console.log(precios[i].price, precios[i].date_time)
            proms_.push(trx('price_today').insert(precios[i]))
        }
        await Promise.all(proms_)
        
        return await trx.commit()
    } else
        return trx.rollback()
}

const PROC_TIME = 70

setTimeout( async ()=>{
    
    await knex('price_today').delete() 

    let productos = await knex('products')
    if (productos){
        console.log('productos obtenidos', productos.length)
        setInterval( async ()=>{
            let producto = productos.pop()
            return await definir_ultimo(producto) 
        }, PROC_TIME)
        
    }

}, 100)

